src/index.js

```diff
@@ -110,39 +120,94 @@ exports.createFromB58String = function (str) {
 }
 
 // Public Key input will be a buffer
-exports.createFromPubKey = function (key) {
+exports.createFromPubKey = function (key, callback) {
   let buf = key
   if (typeof buf === 'string') {
     buf = new Buffer(key, 'base64')
   }
+
+  if (typeof callback !== 'function') {
+    throw new Error('callback is required')
+  }
+
   const pubKey = crypto.unmarshalPublicKey(buf)
-  return new PeerId(pubKey.hash(), null, pubKey)
+  pubKey.hash((err, digest) => {
+    if (err) {
+      return callback(err)
+    }
+
+    callback(null, new PeerId(digest, null, pubKey))
+  })
 }
 
 // Private key input will be a string
-exports.createFromPrivKey = function (key) {
+exports.createFromPrivKey = function (key, callback) {
   let buf = key
   if (typeof buf === 'string') {
     buf = new Buffer(key, 'base64')
   }
 
-  const privKey = crypto.unmarshalPrivateKey(buf)
-  return new PeerId(privKey.public.hash(), privKey)
-}
+  if (typeof callback !== 'function') {
+    throw new Error('callback is required')
+  }
 
-exports.createFromJSON = function (obj) {
-  let priv
-  let pub
+  waterfall([
+    (cb) => crypto.unmarshalPrivateKey(buf, cb),
+    (privKey, cb) => privKey.public.hash((err, digest) => {
+      cb(err, digest, privKey)
+    })
+  ], (err, digest, privKey) => {
+    if (err) {
+      return callback(err)
+    }
 
-  if (obj.privKey) {
-    priv = crypto.unmarshalPrivateKey(new Buffer(obj.privKey, 'base64'))
-  }
+    callback(null, new PeerId(digest, privKey))
+  })
+}
 
-  if (obj.pubKey) {
-    pub = crypto.unmarshalPublicKey(new Buffer(obj.pubKey, 'base64'))
+exports.createFromJSON = function (obj, callback) {
+  if (typeof callback !== 'function') {
+    throw new Error('callback is required')
   }
 
-  return new PeerId(mh.fromB58String(obj.id), priv, pub)
+  const id = mh.fromB58String(obj.id)
+  const rawPrivKey = obj.privKey && new Buffer(obj.privKey, 'base64')
+  const rawPubKey = obj.pubKey && new Buffer(obj.pubKey, 'base64')
+  const pub = rawPubKey && crypto.unmarshalPublicKey(rawPubKey)
+
+  if (rawPrivKey) {
+    waterfall([
+      (cb) => crypto.unmarshalPrivateKey(rawPrivKey, cb),
+      (priv, cb) => priv.public.hash((err, digest) => {
+        cb(err, digest, priv)
+      }),
```

Comment:
It feels weird to me to use waterfall and then capture the callbacks to call the waterfall callback with arguments in a certain order. It is way more simpler to read to just have the shared state between async operations at the top and then run series through them